core/core.go

```diff
@@ -222,62 +258,235 @@ func NewTextile(ctx context.Context, conf Config) (*Textile, error) {
 
-	if err := t.collections.Sessions.Touch(ctx, session); err != nil {
-		return nil, err
+	sid, ok := common.SessionFromMD(ctx)
+	if ok {
+		ctx = common.NewSessionContext(ctx, sid) // For additional requests
+		if sid == t.gatewaySession {
+			return ctx, nil
+		}
+		session, err := t.co.Sessions.Get(ctx, sid)
+		if err != nil {
+			return nil, status.Error(codes.Unauthenticated, "Invalid session")
+		}
+		if time.Now().After(session.ExpiresAt) {
+			return nil, status.Error(codes.Unauthenticated, "Expired session")
+		}
+		if err := t.co.Sessions.Touch(ctx, session.ID); err != nil {
+			return nil, err
+		}
+		ctx = c.NewSessionContext(ctx, session)
+
+		dev, err := t.co.Accounts.Get(ctx, session.Owner)
+		if err != nil {
+			return nil, status.Error(codes.NotFound, "User not found")
+		}
+		ctx = c.NewDevContext(ctx, dev)
+
+		orgSlug, ok := common.OrgSlugFromMD(ctx)
+		if ok {
+			isMember, err := t.co.Accounts.IsMember(ctx, orgSlug, dev.Key)
+			if err != nil {
+				return nil, err
+			}
+			if !isMember {
+				return nil, status.Error(codes.PermissionDenied, "User is not an org member")
+			} else {
+				org, err := t.co.Accounts.GetByUsername(ctx, orgSlug)
+				if err != nil {
+					return nil, status.Error(codes.NotFound, "Org not found")
+				}
+				ctx = c.NewOrgContext(ctx, org)
+				ctx = common.NewOrgSlugContext(ctx, orgSlug) // For additional requests
+				ctx = thread.NewTokenContext(ctx, org.Token)
+			}
+		} else {
+			ctx = thread.NewTokenContext(ctx, dev.Token)
+		}
+	} else if key, ok := common.APIKeyFromMD(ctx); ok {
+		ctx = common.NewAPIKeyContext(ctx, key)
+		msg, sig, ok := common.APISigFromMD(ctx)
+		if !ok {
+			return nil, status.Error(codes.Unauthenticated, "API key signature required")
+		}
+		ctx = common.NewAPISigContext(ctx, msg, sig)
+	} else {
+		return nil, status.Error(codes.Unauthenticated, "Session or API key required")
 	}
+	return ctx, nil
+}
+
+// threadInterceptor monitors for thread creation and deletion.
+// Textile tracks threads against dev, org, and user accounts.
+// Users must supply a valid API key from a dev/org.
+func (t *Textile) threadInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method, _ := grpc.Method(ctx)
+		var isUser bool
+		var owner crypto.PubKey
+		if org, ok := c.OrgFromContext(ctx); ok {
+			owner = org.Key
+		} else if dev, ok := c.DevFromContext(ctx); ok {
+			owner = dev.Key
+		} else if k, ok := common.APIKeyFromContext(ctx); ok {
+			key, err := t.co.Keys.Get(ctx, k)
+			if err != nil || !key.Valid {
+				return nil, status.Error(codes.NotFound, "API key not found or is invalid")
+			}
+			if !common.ValidateAPISigContext(ctx, key.Secret) {
+				return nil, status.Error(codes.PermissionDenied, "Bad API key signature")
+			}
+
+			// Pull user key out of the thread token if present
+			token, ok := thread.TokenFromContext(ctx)
```

Comment:
If a token is present here, it was supplied directly by the user, and we can pull out their user ID (public key)